### PR TITLE
Disable warning for insecure CA if located on private network.

### PR DESCRIPTION
**Pull request by cez81**
My first Caddy PR :)

Fix for #1592. It doesn't handle hostnames that resolves to private IPs.

### CODE REVIEW

**Comment by francislavoie** [comment]
@mholt it helps to simplify IPv6, otherwise we'd need separate logic for IPv6 than IPv4. The address spaces are listed here: https://en.wikipedia.org/wiki/Private_network, for reference

**Comment by cez81** [comment]
Checking the 172.16. -> 172.31. are all internal. But yh could be done with a number of ifs

File: caddy_test.go [hunk 1]
```diff
@@ -61,6 +61,48 @@ func TestIsLoopback(t *testing.T) {
...
+		input  string
+		expect bool
+	}{
+		{"10.0.0.0", true},
+		{"10.0.0.5", true},
+		{"10.255.255.255", true},
+		{"10.0.0.5:1234", true},
+		{"172.16.0.0", true},
+		{"172.19.1.1", true},
+		{"172.31.255.255", true},
```
Comment (mholt) [inline caddy_test.go]: Add tests for the edge cases on both sides of the boundary (both `true` and `false` sides).

File: caddy_test.go [hunk 2]
```diff
@@ -61,6 +61,48 @@ func TestIsLoopback(t *testing.T) {
...
+		{"fd12:3456:789a:1::1:1234", true},
+		{"fd3e:00c3:62a6:e2b3::", true},
+		{"fdff:ffff:ffff:ffff:ffff:ffff:ffff:ffff", true},
+		{"example.com", false},
+		{"localhost", false},
+		{"localhost:1234", false},
+		{"localhost:", false},
+		{"127.0.0.1", false},
+		{"127.0.0.1:443", false},
+		{"127.0.1.5", false},
```
Comment (mholt) [inline caddy_test.go]: Localhost and 127.* are not a private/internal network? I am thinking I disagree...

Comment (francislavoie) [inline caddy_test.go]: Technically, loopback is different than private. We already have `IsLoopback` here: https://github.com/cez81/caddy/blob/1f928fc764da0b88ffa5117bdd2853aa20bd4446/caddy.go#L756 so we can just do `IsLoopback || IsPrivateNetwork` depending on the use-case.

Comment (mholt) [inline caddy_test.go]: Except I feel like private/internal is a _superset_ of loopback... in other words, almost feels like we should add a check for IsLoopback into this function.

Comment (francislavoie) [inline caddy_test.go]: Maybe this is a weird idea, but maybe we make this function `IsPrivate` (for naming consistency, we're not calling the other `IsLoopbackNetwork`) and make another helper called `IsPrivateNetwork` which gives us the union. I think there's value in having `IsPrivate` its own thing for potential future use.

Comment (mholt) [inline caddy_test.go]: Well, both are for networks, so how about this instead:

- `IsLoopback` (keep)
- `IsInternal` (rename this new function)
- and then we just use `IsLoopback() && IsInternal()` when we need both.

Make sure the godoc comment for IsInternal states that it does not include loopback addresses.

Comment (francislavoie) [inline caddy_test.go]: It would be `||` not `&&`, it can never be both loopback and internal 😉 
Sounds good to me.

**Comment by mholt** [comment]
Looks good, @cez81! Thanks for contributing, and nice work!
